Fix bootchooser for the output of efibootmgr 18

dvzrv opened the pull request:
Since efibootmgr 18, the default output of `efibootmgr` is now more verbose [1], which breaks the assumptions made by RAUC in regards to parsing the output. This issue is also affecting others [2].

Fix the parsing logic by unconditionally splitting on a tab in the detected name part of the boot entry, so that the unused data part can be discarded.
Emit a debug message for each found boot entry, as this helps in detecting issues in the future.

[1] https://github.com/rhboot/efibootmgr/commit/8ec3e9dedb3cb62f19847794012420b90f475398
[2] https://github.com/rhboot/efibootmgr/issues/169

<!--
Thank you for your pull request!

If this is your first pull request for RAUC, please read:
https://rauc.readthedocs.io/en/latest/contributing.html

Please describe what it changes, e.g. fixes a bug (and how), adds a feature, fixes documentation…

If you add a feature, please answer these questions:
- What do you use the feature for?
- How does RAUC benefit from the feature?
- How did you verify the feature works?
- If hardware is needed for the feature, which hardware is supported and which
  hardware did you test with?

Please also allow edits by maintainers, so we can apply minor fixes directly.
-->

<!--
In case your PR fixes an issue, please reference it in the next line, i.e.
Fixes: #[insert number without brackets here]
-->

This fix is not related to an open ticket, as I noticed this quite recently and got in touch on matrix to discuss a possible fix.
I verified this to work in an Arch Linux VM using efibootmgr 18 and a RAUC 1.10 with this patch applied.

As writing C is not really my forte, I'd appreciate feedback (especially whether this works as intended with efibootmgr < 18).

jluebbe commented on src/bootchooser.c:
```diff
@@ -1159,11 +1159,26 @@ static gboolean efi_bootorder_get(GList **bootorder_entries, GList **all_entries
 		efi_bootentry *entry = g_new0(efi_bootentry, 1);
 		entry->num = g_match_info_fetch(match, 1);
-		entry->name = g_match_info_fetch(match, 2);
+		entry->name = g_strdup(name_data_splits[0]);
```
Instead of using `g_strsplit`, you can also modify entry->name in place. Use strchr to find the first tab, and if there is one, replace it by `'\0'`. This also avoids any additional memory allocations.

dvzrv replied:
Ah good point. I'll try that... might take me some time though.
Better than to leak memory.. phew :sweat_smile:

dvzrv replied:
I'll try with cleaning up my current code further, as the glib provided functions are easier for me to use right now.
I'd be happy to see whether with the recent change I am circumventing the memory leak already.

jluebbe reviewed (changes requested):
To make sure this can work reliably, we need to check that the boot name doesn't contain whitespace. See https://github.com/rauc/rauc/blob/master/src/config_file.c#L235 for the place where it is read. That area also has examples for reporting an error. The bootname check should be in a helper function.

dvzrv commented:
> To make sure this can work reliably, we need to check that the boot name doesn't contain whitespace. See https://github.com/rauc/rauc/blob/master/src/config_file.c#L235 for the place where it is read. That area also has examples for reporting an error. The bootname check should be in a helper function.

Maybe I lack context here, but why can it not contain a whitespace? This seems certainly possible in the context of e.g. `efibootmgr`.

jluebbe commented:
> > To make sure this can work reliably, we need to check that the boot name doesn't contain whitespace. See https://github.com/rauc/rauc/blob/master/src/config_file.c#L235 for the place where it is read. That area also has examples for reporting an error. The bootname check should be in a helper function.
> 
> Maybe I lack context here, but why can it not contain a whitespace? This seems certainly possible in the context of e.g. `efibootmgr`.

We currently don't have checks for the bootname, so it could contain a tab character. In that case, applying this change would break at runtime. Adding a check during config parsing would report that issue early. And while adding that check, we should also reject any unusual bootnames containing whitespace (to be better prepared for future efibootmgr changes).

dvzrv commented:
> We currently don't have checks for the bootname, so it could contain a tab character. In that case, applying this change would break at runtime. Adding a check during config parsing would report that issue early. And while adding that check, we should also reject any unusual bootnames containing whitespace (to be better prepared for future efibootmgr changes).

I see. But validating the config input and more robustly dealing with the output of efibootmgr are two separate issues (in my mind).
Are there any straightforward facilities to search for the right-most tab character? Maybe that would make more sense.

jluebbe commented:
I've adjusted the commit titles to our usual format, made some changes to the code (see the fixup commits) and added a test-case for the config value check. @dvzrv does this look OK for you? Could you test it with efibootmgr 18?

dvzrv commented:
> I've adjusted the commit titles to our usual format, made some changes to the code (see the fixup commits) and added a test-case for the config value check. @dvzrv does this look OK for you? Could you test it with efibootmgr 18?

Thanks! I've rebuild the package for Arch Linux locally with the updated fixes. Tests run fine and after installing the updated package in a VM (with A/B setup) the output of `rauc status --detailed` works without issue.

jluebbe commented:
Spaces/tabs are also problematic for bootnames in other places, like the kernel cmdline, so it's better to reject them early.

The strchr approach seems simpler, so I prefer it for the point release.

ejoerns commented:
> Spaces/tabs are also problematic for bootnames in other places, like the kernel cmdline, so it's better to reject them early.

Since a broken config will fail early with this approach, go for it.

> The strchr approach seems simpler, so I prefer it for the point release.

Agreed for the point release.
